test-data/unit/check-classes.test

```diff
@@ -894,6 +892,281 @@ a.f = a.f # E: Property "f" defined in "A" is read-only
+
+[case testAccessingNonDataDescriptorSubclass]
+class C():
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+class D(C): pass
+
+class A:
+    f = D()
+
+a = A() # type: A
+reveal_type(a.f)  # E: Revealed type is 'builtins.str'
+
+[case testSettingDataDescriptorSubclass]
+class C():
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class D(C): pass
+
+class A:
+    f = D()
+
+a = A() # type: A
+a.f = ''
+a.f = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+[case testReadingDescriptorSubclassWithoutDunderGet]
+from typing import Union
+
+class C():
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class D(C): pass
+
+class A:
+    f = D()  # type: Union[D, str]
+
+    def __init__(self):
+        self.f = 'some string'  # type: str
+
+a = A() # type: A
+reveal_type(a.f)  # E: Revealed type is 'Union[__main__.D, builtins.str]'
+
+[case testAccessingGenericNonDataDescriptorSubclass]
+from typing import TypeVar, Type, Generic
+
+T = TypeVar('T')
+V = TypeVar('V')
+class C(Generic[T, V]):
+    def __init__(self, value: V) -> None:
+        self.value = value
+    def __get__(self, instance: T, owner: Type[T]) -> V:
+        return self.value
+
+class D(C[T, V], Generic[T, V]): pass
+
+class A:
+    f = D(10)  # type: D[A, int]
+    g = D('10')  # type: D[A, str]
+
+a = A() # type: A
+reveal_type(a.f)  # E: Revealed type is 'builtins.int*'
+reveal_type(a.g)  # E: Revealed type is 'builtins.str*'
+
+[case testSettingGenericDataDescriptorSubclass]
+from typing import TypeVar, Type, Generic
+
+T = TypeVar('T')
+V = TypeVar('V')
+class C(Generic[T, V]):
+    def __init__(self, value: V) -> None:
+        self.value = value
+    def __get__(self, instance: T, owner: Type[T]) -> V:
+        return self.value
+    def __set__(self, instance: T, value: V) -> None:
+        pass
+
+class D(C[T, V], Generic[T, V]): pass
+
+class A:
+    f = D(10)  # type: D[A, int]
+    g = D('10')  # type: D[A, str]
+
+a = A() # type: A
+a.f = 1
+a.f = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+a.g = ''
+a.g = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+[case testSetDescriptorOnClass]
+from typing import TypeVar, Type, Generic
+
+T = TypeVar('T')
+V = TypeVar('V')
+class D(Generic[T, V]):
+    def __init__(self, value: V) -> None:
+        self.value = value
+    def __get__(self, instance: T, owner: Type[T]) -> V:
+        return self.value
+    def __set__(self, instance: T, value: V) -> None:
+        pass
+
+class A:
+    f = D(10)  # type: D[A, int]
+
+A.f = D(20)
+A.f = D('some string')  # E: Argument 1 to "D" has incompatible type "str"; expected "int"
+
+
+-- Multiple inheritance, non-object built-in class as base
```

Comment:
The next two tests were actually deleted. I guess you accidentally restored them as part of a merge resolution?